File: files-app/src/main/java/l/files/operations/Count.java
package l.files.operations;

import java.io.IOException;
import java.util.Collection;
import java.util.concurrent.atomic.AtomicInteger;

import l.files.fs.File;

class Count extends AbstractOperation {

  private final AtomicInteger count = new AtomicInteger();

  Count(Collection<? extends File> resources) {
    super(resources);
  }

  public int getCount() {
    return count.get();
  }

  @Override void process(File file) {
    traverse(file, new OperationVisitor() {

      @Override public Result onPreVisit(File res) throws IOException {
        count.incrementAndGet();
        onCount(res);
        return super.onPreVisit(res);
      }

    });
  }

  void onCount(File file) {
  }

}

File: files-app/src/main/kotlin/operations/Count.kt
package l.files.operations

import java.io.IOException

import l.files.fs.Path
import l.files.fs.Resource

import l.files.fs.Resource.TraversalOrder.BREATH_FIRST

private open class Count(paths: Iterable<Path>) : AbstractOperation(paths) {

    volatile var count = 0
        private set

    override fun process(path: Path, listener: FailureRecorder) {
        try {
            count(path, listener)
        } catch (e: IOException) {
            listener.onFailure(path, e)
        }
    }

    private fun count(path: Path, listener: FailureRecorder) {
        path.resource
                .traverse(BREATH_FIRST) { res, err -> listener.onFailure(res.path, err) }
                .takeWhile { !Thread.currentThread().isInterrupted() }
                .forEach {
                    count++
                    onCount(it)
                }
    }

    open fun onCount(resource: Resource) {
    }

}
